src/coreComponents/constitutive/fluid/BlackOilFluid.cpp

```diff
@@ -31,76 +26,443 @@ using namespace dataRepository;
+  {
+    real64 const dPext = m_PVTO.m_maxRelativePressure - m_PVTO.m_undersaturatedPressure[i].back();
+
+    array1d< real64 > const & Pusat = m_PVTO.m_undersaturatedPressure[i];
+    array1d< real64 > const & Bousat = m_PVTO.m_undersaturatedBo[i];
+    array1d< real64 > const & viscusat = m_PVTO.m_undersaturatedViscosity[i];
+
+    if( std::fabs( dPext ) > 0 )
+    {
+      localIndex const branchSize = m_PVTO.m_undersaturatedPressure[i].size();
+      real64 const Bo = linearExtrapolation( Pusat[branchSize - 2],
+                                             Bousat[branchSize - 2],
+                                             Pusat[branchSize - 1],
+                                             Bousat[branchSize - 1],
+                                             m_PVTO.m_maxRelativePressure );
+      real64 const visc = linearExtrapolation( Pusat[branchSize - 2],
+                                               viscusat[branchSize - 2],
+                                               Pusat[branchSize - 1],
+                                               viscusat[branchSize - 1],
+                                               m_PVTO.m_maxRelativePressure );
+      m_PVTO.m_undersaturatedBo[i].emplace_back( Bo );
+      m_PVTO.m_undersaturatedViscosity[i].emplace_back( visc );
+      m_PVTO.m_undersaturatedPressure[i].emplace_back( m_PVTO.m_maxRelativePressure );
+    }
+  }
 }
 
-void BlackOilFluid::postProcessInput()
+void BlackOilFluid::refineTableAndCopy( localIndex const nLevels )
 {
-  // TODO maybe use different names?
-  m_componentNames = m_phaseNames;
+  std::vector< real64 > refineP; // TODO:change to array1d
+  for( localIndex i = 0; i < m_PVTO.m_nSaturatedPoints; ++i )
+  {
+    for( localIndex j = 0; j < m_PVTO.m_undersaturatedPressure[i].size(); ++j )
+    {
+      refineP.emplace_back( m_PVTO.m_undersaturatedPressure[i][j] );
+    }
+  }
+  std::sort( refineP.begin(), refineP.end() );
+  refineP.erase( std::unique( refineP.begin(), refineP.end() ), refineP.end() );
+
+  // linspace
+  real64 const maxVal = *std::max_element( refineP.begin(), refineP.end() );
+  array1d< real64 > refinePP( refineP.size() );
+  for( localIndex i = 0; i < localIndex( refineP.size() ); ++i )
+  {
+    refinePP[i] = refineP[i];
+  }
+
+  refinePP.resize( nLevels );
+  for( localIndex i = 0; i != nLevels; ++i )
+  {
+    refinePP[i] = i * ( maxVal / ( nLevels - 1 ) );
+  }
 
-  MultiFluidPVTPackageWrapper::postProcessInput();
+  localIndex const ppSize = refinePP.size();
+  array1d< real64 > propRefined( ppSize );
+  for( localIndex i = 0; i < m_PVTO.m_nSaturatedPoints; ++i )
+  {
 
-  localIndex const NP = numFluidPhases();
+    // Bo
+    interpolation1( m_PVTO.m_undersaturatedPressure[i], m_PVTO.m_undersaturatedBo[i], refinePP, propRefined );
+    m_PVTO.m_undersaturatedBo[i].resize( ppSize );
+    for( localIndex j = 0; j < ppSize; ++j )
+    {
+      m_PVTO.m_undersaturatedBo[i][j] = propRefined[j];
+    }
```

Comment:
I think loops like this can be replaced with just array assignment, i.e.
```
m_PVTO.m_undersaturatedBo[i] = propRefined;
```
or even using `m_PVTO.m_undersaturatedBo[i]` directly as the argument to `interpolation1`